### DISCUSSION TITLE
HTTP 상태코드 및  메소드

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#96] [프론트 기술 질문 공간]
HTTP 상태코드 및  메소드에 대해 얼마나 알고 계시나요?

ex) 1xx 코드는 어떤 의미인가요?

ex) get 메소드와 post 메소드의 차이는 무엇인가요?

**Comment by michoball** [comment]
HTTP 메서드의 속성으로는 크게 3 가지인 안전(Safe), 멱등(Idempotent), 캐시 가능(Cacheable)이 있다.

안전성: 보안 취약성을 말하는 것이 아니라 호출해도 리소스가 변경되지 않는 성질을 말하는 것

멱등성:  요청(Request)을 한 번을 호출하든 여러 번을 호출하든 그 결과가 같음을 의미 여기서 결과가 의미하는 것은 응답 상태 코드가 아닌 서버의 상태

캐시 가능:  응답 결과 리소스를 캐싱해서 효율적으로 사용할 수 있는가를 말함


 메소드들

GET: 데이터를 요청할때 사용 ( 데이터를 포함하고 있으면 안됨 )

POST:  데이터를 서버로 보낼 때 사용 Content-Type으로 데이터 타입을 지정할 수 있다.  put과의 다른점은 성공적인 요청 이후에 부수적인 효과가 있다는 것

PUT:  새로운 데이터를 생성하거나 지정한 데이터를 수정할 때 쓰임 

DELETE:  특정 리소스 삭제 요청 

PATCH:  PUT처럼 데이터를 수정하나 부분적으로만 수정할때 사용 (업데이트 컨)

OPTIONS:  서버와 통신하기위해 통신 옵션을 확인하기 위해 사용된다. 응답받은 값을 보고 이후의 요청이 가능한지 파악한다. (prefilght 요청으로 쓰임)

CONNECT:  two-way 소통을 위해 사용되는 메서드, CONNECT 메소드는 SSL(HTTPS)를 사용하는 웹 사이트 접속하는데 사용할 수 있다.  클라이언트는 원하는 목적지와 TCP연결을 HTTP 프록시 서버에 요청한다. 그러면 서버는 클라이언트 대신하여 연결의 생성을 진행한다. 한번 서버에 의해 연결이 수립되면, 프록시 서버는 클라이언트에 오고가는 TCP 스트림을 계속해서 프록시한다 

HEAD:  GET과 유사하지만 실제 문서가 아닌  header 의 정보를 얻을 때 사용 

TRACE:  요청 리소스가 수신되는 경로를 보여줌.

상태코드: 백번때만 

1XX : 서버가 요청을 받았으며 서버에 연결된 클라이언트는 작업을 계속 진행하라는 의미 

2XX : 요청이 성공했을때 코드

3XX :  요청 완료를 위해 추가 작업조치가 필요하다는 의미
클라이언트가 요청한 리소스가 옮겨졌거나 리소스가 삭제되었거나해서 정상적인 방법으로는 더 이상 해당 리소스에 접근할 수 없고 다른 URL을 통해서 그 리소스에 접근해야하는 경우 서버는 “여기로 가면 니가 찾는 리소스가 있어!”라는 정보를 알려줄 수 있는데, 이때 사용되는 상태 코드들이 바로 300번대 코드들이다.

4xx :  클라이언트 오류 - 요청 문법이 잘못되었거나 요청을 처리할 수 없다는 의미

5xx : 서버가 명백히 유효한 요청에 대한 충족을 실패했다는 의미

**Comment by WooGie911** [comment]
![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/14618dac-b992-4ff2-a6d6-ff11e4119259)

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/571739d2-a591-4f52-b107-521e379eef60)


### PUT vs POST
- PUT은 서버의 Resource에 데이터를 저장하기 위한 용도
- POST는 서버에 데이터를 보내기 위한 용도

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/8403fff0-ab7e-455b-a8a6-39526fb00e2b)



## HTTP Status Code

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/d43fde85-a913-4df0-b1ad-ac2450cf2ec9)


![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/87591d95-2ac1-4edb-9d38-bdaad70a306b)


![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/e1e39891-a56e-487a-8749-3e73c75af57a)


![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/31ed9c3a-dda2-404d-84fa-e867ddc26620)

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/7c1aee66-e103-4b75-a11c-87bf4f768f44)


![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/ca8a29e8-fa54-4be3-8b12-612e58bf2b58)


![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/ba54e8ad-f82b-47bc-8175-65f274fecd29)

**Comment by javascriptbiggosoo** [comment]
### 메서드
<html>
<body>
<p>여러 종류의 request 명령을 HTTP 메서드로 구분해놓았습니다.(client → server)</p>
<p>메서드는 서버가 어떤 동작을 취해야 하는지 알려줍니다.(웹페이지 가져오기, 게이트웨이 프로그램 실행하기, 파일 삭제하기 등)</p>
<p>하나의 HTTP request는 하나의 메서드를 갖습니다.</p>


HTTP method | Description
-- | --
GET | Send named resource from the server to the client.
PUT | Store data from client into a named server resource.
DELETE | Delete the named resource from a server.
POST | Send client data into a server gateway application.
HEAD | Send just the HTTP headers from the response for the named resource.






</body>
</html>

### Status Codes
<html>
<body>
<p>모든 HTTP response 메시지는 status code(상태 코드)와 함께 반환됩니다.(server → client)</p>
<p>클라이언트는 상태 코드 세자리 숫자를 통해 요청이 성공했는지, 추가 조치가 필요한지 알 수 있습니다.</p>
<p>아래는 가장 흔히 쓰이는 세 가지 상태코드 입니다.</p>


HTTP status code | Description
-- | --
200 | OK. Document returned correctly.
302 | Redirect. Go someplace else to get the resource.
404 | Not Found. Can’t find this resource.


<p>숫자로 된 status code와 더불어 왜 status가 이 꼴이 났는지를 설명해주는 텍스트인 “reson phrase” 도 함께 보내는데, reason phrase는 사람에게 설명하기 위한 것이며 실질적인 response의 포로세싱에는 status code만 사용됩니다.</p>

</body>
</html>
